python/lsst/daf/butler/registry/_registry.py

```diff
@@ -287,109 +264,23 @@ def defaults(self, value: RegistryDefaults) -> None:
         value.finish(self)
         self._defaults = value
 
+    @abstractmethod
     def refresh(self) -> None:
         """Refresh all in-memory state by querying the database.
 
         This may be necessary to enable querying for entities added by other
-        `Registry` instances after this one was constructed.
+        registry instances after this one was constructed.
         """
-        self._managers.refresh()
+        raise NotImplementedError()
 
     @contextlib.contextmanager
+    @abstractmethod
     def transaction(self, *, savepoint: bool = False) -> Iterator[None]:
         """Return a context manager that represents a transaction.
         """
-        try:
-            with self._db.transaction(savepoint=savepoint):
-                yield
-        except BaseException:
-            # TODO: this clears the caches sometimes when we wouldn't actually
-            # need to.  Can we avoid that?
-            self._managers.dimensions.clearCaches()
-            raise
-
-    def resetConnectionPool(self) -> None:
```

Comment:
This is probably SQL-specific, but removing it from interface will force clients to check the actual type of the Registry before calling it. Maybe better to keep it but provide default implementation that does nothing?

Reply:
Okay. mypy didn't complain so I ignored it. I had not noticed that ctrl_mpexec was relying on it.